### ISSUE TITLE
Performance Decline and Data Preprocessing Query for CausalARFlow Model

### ISSUE THREAD

**Issue by hyxie2023** [rgklab/StructuredNNs#7]
Hello, thank you for open-sourcing this excellent work!

I've been trying to run the run_experiment file with default parameters and achieved the expected results. However, when I increase the coefficient of SEM, the performance of the model significantly declines. For instance, when I modify the `adj_gen_param` of [SparseSEM](https://github.com/rgklab/StructuredNNs/blob/main/data/causal_sem.py) from the default `(-2,2)` to `(0,3)` while keeping other parameters constant, the log_likelihood of test sets carefl and straf respectively becomes -9.91 and -9.61, and the results of interventions and counterfactuals testing are also unsatisfactory.

I'd like to inquire about whether [CausalARFlow](https://github.com/rgklab/StructuredNNs/blob/main/strnn/models/causal_arflow.py) imposes any requirements on the distribution of input data and how I should preprocess the data to better suit the model's handling. Thank you!

**Comment by IanShi1996** [comment]
Thanks for bringing this to our attention, and apologies for the delayed response!

Changing the `adj_gen_param` to `(0, 3)` will generate a SEM with only positive coefficients -- consequently, the magnitude of the variables to increase further in the causal ordering. After some testing, it seems that Causal StrAF tends to "blow up" in response to high variable magnitudes. Empirically, increasing the weight decay used during training seems to mitigate this effect.

For example, on the default graph seed with `adj_gen_param = (0, 3)` and 500 data samples, training Causal StrAF with no weight decay results in an I-MSE 2906.86 while adding weight decay with a value of 0.005 results in an I-MSE of 75.24. Note that this performance is without any hyper-parameter tuning, so you'll likely have to play around with the weight decay and other HPs to get a good fit on the specific dataset instead of using the default values. We'll be adding some code to early stop the Causal StrAF during training to help with this process. 

Hope this helps!

**Comment by hyxie2023** [comment]
Thank you so much for your prompt and detailed response! Your insights are incredibly helpful and have given me a lot to think about. StrAF is a powerful tool, and I'm currently exploring ways to incorporate your innovative approach into downstream tasks I'm working on.

I'll definitely keep an eye on your updates on GitHub. Looking forward to more exchanges and learning from your work in the future.

Best regards,
Hongyi Xie

在2024-03-21 07:38:55，Ian ***@***.***写道：

Thanks for bringing this to our attention, and apologies for the delayed response!

Changing the adj_gen_param to (0, 3) will generate a SEM with only positive coefficients -- consequently, the magnitude of the variables to increase further in the causal ordering. After some testing, it seems that Causal StrAF tends to "blow up" in response to high variable magnitudes. Empirically, increasing the weight decay used during training seems to mitigate this effect.

For example, on the default graph seed with adj_gen_param = (0, 3) and 500 data samples, training Causal StrAF with no weight decay results in an I-MSE 2906.86 while adding weight decay with a value of 0.005 results in an I-MSE of 75.24. Note that this performance is without any hyper-parameter tuning, so you'll likely have to play around with the weight decay and other HPs to get a good fit on the specific dataset instead of using the default values. We'll be adding some code to early stop the Causal StrAF during training to help with this process.

Hope this helps!
